diff --git dropbear/dbmulti.c dropbear/dbmulti.c
index 3299970..1e5f8d7 100644
--- dropbear/dbmulti.c
+++ dropbear/dbmulti.c
@@ -55,11 +55,13 @@ static int runprog(const char *progname, int argc, char ** argv, int *match) {
 			return dropbearconvert_main(argc, argv);
 		}
 #endif
-#if 0//def DBMULTI_scp
+#ifdef HAVE_TOMATO
+#ifdef DBMULTI_scp
 		if (strcmp(progname, "scp") == 0) {
 			return scp_main(argc, argv);
 		}
 #endif
+#endif /* HAVE_TOMATO */
 	*match = DROPBEAR_FAILURE;
 	return 1;
 }
@@ -94,9 +96,11 @@ int main(int argc, char ** argv) {
 #ifdef DBMULTI_dropbearconvert
 			"'dropbearconvert' - the key converter\n"
 #endif
-#if 0//def DBMULTI_scp
+#ifdef HAVE_TOMATO
+#ifdef DBMULTI_scp
 			"'scp' - secure copy\n"
 #endif
+#endif /* HAVE_TOMATO */
 			,
 			DROPBEAR_VERSION);
 	exit(1);
