diff --git dropbear/dbmulti.c dropbear/dbmulti.c
index 1bc70ea..78704c3 100644
--- dropbear/dbmulti.c
+++ dropbear/dbmulti.c
@@ -50,11 +50,13 @@ static int runprog(const char *multipath,
 			return dropbearconvert_main(argc, argv);
 		}
 #endif
-#if 0//def DBMULTI_scp
+#ifdef HAVE_TOMATO
+#ifdef DBMULTI_scp
 		if (strcmp(progname, "scp") == 0) {
 			return scp_main(argc, argv);
 		}
 #endif
+#endif /* HAVE_TOMATO */
 	*match = DROPBEAR_FAILURE;
 	return 1;
 }
@@ -93,9 +95,11 @@ int main(int argc, char ** argv) {
 #ifdef DBMULTI_dropbearconvert
 			"'dropbearconvert' - the key converter\n"
 #endif
-#if 0//def DBMULTI_scp
+#ifdef HAVE_TOMATO
+#ifdef DBMULTI_scp
 			"'scp' - secure copy\n"
 #endif
+#endif /* HAVE_TOMATO */
 			,
 			DROPBEAR_VERSION);
 	exit(1);
